File: sdk/ios/Sources/MessageBirdsAnalytics/EventEnvelope.swift
import Foundation

/// Mirrors `mb_core::IdentityRef` (see `sdk/js/src/types.ts`, `sdk/python/src/messagebirds/types.py`).
public struct IdentityRef: Codable, Equatable {
    public let namespace: String
    public let value: String
    public let primary: Bool?
    public let source: String
    public let confidence: Double?

    public init(namespace: String, value: String, primary: Bool? = nil, source: String, confidence: Double? = nil) {
        self.namespace = namespace
        self.value = value
        self.primary = primary
        self.source = source
        self.confidence = confidence
    }
}

/// Mirrors `mb_core::EventSource`.
public struct EventSource: Codable, Equatable {
    public let type: String
    public let name: String

    public init(type: String, name: String) {
        self.type = type
        self.name = name
    }
}

/// Mirrors `mb_core::SchemaRef`.
public struct SchemaRef: Codable, Equatable {
    public let name: String
    public let version: String

    public init(name: String, version: String) {
        self.name = name
        self.version = version
    }
}

/// A JSON value with no fixed shape — used for `data`/`context`, matching
/// `Record<string, unknown>` on the TS side and `dict[str, Any]` on the
/// Python side.
public enum JSONValue: Codable, Equatable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case object([String: JSONValue])
    case array([JSONValue])
    case null

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let v = try? container.decode(Bool.self) {
            self = .bool(v)
        } else if let v = try? container.decode(Double.self) {
            self = .number(v)
        } else if let v = try? container.decode(String.self) {
            self = .string(v)
        } else if let v = try? container.decode([String: JSONValue].self) {
            self = .object(v)
        } else if let v = try? container.decode([JSONValue].self) {
            self = .array(v)
        } else {
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "Unsupported JSON value")
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let v): try container.encode(v)
        case .number(let v): try container.encode(v)
        case .bool(let v): try container.encode(v)
        case .object(let v): try container.encode(v)
        case .array(let v): try container.encode(v)
        case .null: try container.encodeNil()
        }
    }
}

/// Mirrors `mb_core::EventEnvelope` — the JSON body `POST /events` expects.
public struct EventEnvelope: Codable, Equatable {
    public let event_id: String
    public let event_type: String
    public let timestamp: String
    public let source: EventSource
    public let identity: [IdentityRef]
    public let context: [String: JSONValue]?
    public let data: [String: JSONValue]?
    public let schema: SchemaRef
    public let tenant_id: String

    public init(
        event_id: String,
        event_type: String,
        timestamp: String,
        source: EventSource,
        identity: [IdentityRef],
        context: [String: JSONValue]? = nil,
        data: [String: JSONValue]? = nil,
        schema: SchemaRef,
        tenant_id: String
    ) {
        self.event_id = event_id
        self.event_type = event_type
        self.timestamp = timestamp
        self.source = source
        self.identity = identity
        self.context = context
        self.data = data
        self.schema = schema
        self.tenant_id = tenant_id
    }
}

File: sdk/android/src/main/kotlin/com/messagebirds/android/EventEnvelope.kt
package com.messagebirds.android

import org.json.JSONArray
import org.json.JSONObject

/** Mirrors `mb_core::IdentityRef` (see `sdk/js/src/types.ts`, `sdk/ios/.../EventEnvelope.swift`). */
data class IdentityRef(
    val namespace: String,
    val value: String,
    val primary: Boolean? = null,
    val source: String,
    val confidence: Double? = null,
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("namespace", namespace)
        put("value", value)
        primary?.let { put("primary", it) }
        put("source", source)
        confidence?.let { put("confidence", it) }
    }
}

/** Mirrors `mb_core::EventSource`. */
data class EventSource(val type: String, val name: String) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("type", type)
        put("name", name)
    }
}

/** Mirrors `mb_core::SchemaRef`. */
data class SchemaRef(val name: String, val version: String) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("name", name)
        put("version", version)
    }
}

/**
 * Mirrors `mb_core::EventEnvelope` — the JSON body `POST /events` expects.
 * `data`/`context` are `org.json.JSONObject` directly (Android's built-in
 * JSON support) rather than a generic map + a serialization dependency —
 * callers building nested structures use `JSONObject`/`JSONArray` the same
 * way they would for any other Android networking code.
 */
data class EventEnvelope(
    val eventId: String,
    val eventType: String,
    val timestamp: String,
    val source: EventSource,
    val identity: List<IdentityRef>,
    val context: JSONObject? = null,
    val data: JSONObject? = null,
    val schema: SchemaRef,
    val tenantId: String,
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("event_id", eventId)
        put("event_type", eventType)
        put("timestamp", timestamp)
        put("source", source.toJson())
        put("identity", JSONArray(identity.map { it.toJson() }))
        context?.let { put("context", it) }
        data?.let { put("data", it) }
        put("schema", schema.toJson())
        put("tenant_id", tenantId)
    }

    companion object {
        fun fromJson(json: JSONObject): EventEnvelope {
            val identityArray = json.getJSONArray("identity")
            val identity = (0 until identityArray.length()).map { i ->
                val obj = identityArray.getJSONObject(i)
                IdentityRef(
                    namespace = obj.getString("namespace"),
                    value = obj.getString("value"),
                    primary = if (obj.has("primary")) obj.getBoolean("primary") else null,
                    source = obj.getString("source"),
                    confidence = if (obj.has("confidence")) obj.getDouble("confidence") else null,
                )
            }
            val sourceObj = json.getJSONObject("source")
            val schemaObj = json.getJSONObject("schema")
            return EventEnvelope(
                eventId = json.getString("event_id"),
                eventType = json.getString("event_type"),
                timestamp = json.getString("timestamp"),
                source = EventSource(sourceObj.getString("type"), sourceObj.getString("name")),
                identity = identity,
                context = json.optJSONObject("context"),
                data = json.optJSONObject("data"),
                schema = SchemaRef(schemaObj.getString("name"), schemaObj.getString("version")),
                tenantId = json.getString("tenant_id"),
            )
        }
    }
}
